### ISSUE TITLE
BatchUpdateManager.InMemoryDbContextFactory per instance

### ISSUE THREAD

**Issue by agustins** [zzzprojects/EntityFramework-Plus#344]
`BatchUpdateManager.InMemoryDbContextFactory` is implemented as static. 
This causes that parallel tests or task changes the context running the BatchUpdate to diferents contexts and failing.

It would be great to have an instance configuration of BatchUpdateManager.InMemoryDbContextFactory. 

related to #149

**Comment by JonathanMagnan** [comment]
Hello @agustins ,

Thank you for your suggestion, we will try to look at it this week.

Best Regards,

Jonathan

**Comment by agustins** [comment]
Hello @JonathanMagnan any update for this issue? Thanks!

**Comment by JonathanMagnan** [comment]
Hello @agustins ,

Unfortunately, I don't have any news about it.

We are currently getting 3x more request support as we used to have to we currently try to find our breath ;(

I moved the priority higher on this request since you contacted us, so it should be done sooner. I will try to have my developer to do it on next week.

Best Regards,

Jonathan

**Comment by ahmettahasakar** [comment]
Hello,

Is there any update on this? Thank you

**Comment by agustins** [comment]
Hi @JonathanMagnan we need to know if this could be solved soon, our tests run in parallel and this is very inestable. Sorry for my request but if we cannot solve it we'll change the library. Thanks in advance!

**Comment by JonathanMagnan** [comment]
Hello @agustins ,

I do not understand why this request has not been done yet. I will check with my developer.

It should now be very easy to do such request

**Comment by JonathanMagnan** [comment]
Hello,

Starting from the v3.0.32, it's now possible to set the value of InMemoryDbContextFactory per instance.

```csharp
context.EntitySimple.Delete(options => options.InMemoryDbContextFactory = () => theContextFactory);
context.EntitySimple.Update(x => new EntitySimple() { ColumnInt = 2 }, options => options.InMemoryDbContextFactory = () => theContextFactory)};
```

Please confirm that everything is working as expected.

Best Regards,

Jon

**Comment by agustins** [comment]
@JonathanMagnan yes it works! Thank you very much!

**Comment by kdvtrifork** [comment]
Hi! 

Can this be done without doing `Delete(options => options.InMemoryDbContextFactory = () => theContextFactory)`
Otherwise we have to add test code to out production code? 

Or am I understanding this wrong?

**Comment by JonathanMagnan** [comment]
Hello @kdvtrifork ,

What exactly is your requirement? I'm not sure to understand this part `we have to add test code to our production code`, obviously you should not have to do this!

**Comment by kdvtrifork** [comment]
Thanks for quick answer @JonathanMagnan, 

We have multiple repositories that uses `.DeleteFromQueryAsync()` or `UpdateFromQuery()`. 
So if we use `BatchUpdateManager.InMemoryDbContextFactory` then we get random errors as mentioned in the issue above.

So my question is:
Where do we need to set the `options => options.InMemoryDbContextFactory = () => theContextFactory;` code. 
I don't see where in the unit test class to set it.

**Comment by JonathanMagnan** [comment]
Here is an example of use:

```csharp
var inMemoryOptions = new DbContextOptionsBuilder<EntityContext>()
	.UseInMemoryDatabase(databaseName: "Add_writes_to_database")
	.Options;
				
using (var context = new EntityContext(inMemoryOptions))
{
	context.Customers.DeleteFromQuery(options => options.InMemoryDbContextFactory = () => new EntityContext(inMemoryOptions));
}
```

In your unit test, this really depends on how it has been created. That's pretty hard to answer this one without having your code.

I recommend you to create another Issue since this one is already answered and provide us a runnable example that contains your code, it will be easier for us to provide a good answer
